### PR TITLE
Add License

### PR DESCRIPTION
**Pull request by mayala1925**
Adding License file.

### CODE REVIEW

**Review by jjc2718 — APPROVED** [review]
Looks good!

File: stratified vs. random cross-val.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        
+        # Getting standard deviation of each k-fold section
+        standard_deviation = np.std(accuracy_scores)
+        standard_deviations.append(standard_deviation)
+        
+    return noise_list,mean_acc_scores,standard_deviations
+
+
+# Creating a list that will be used for the different cutoffs in labeling.
+cut_off_list = [0.01,0.25,0.5,0.75,0.99]
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]:
```suggestion
cut_off_list = [0.01, 0.25, 0.5, 0.75, 0.99]
```

File: stratified vs. random cross-val.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+
+# Looping through cutoffs and graphing each model at each labeling breakpoint
+for i in cut_off_list:
+    noise,accuracy,errorbars = get_noise_accuracy(np.arange(0,100,5), np.arange(-10,10,0.01), i)
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]:
```suggestion
    noise, accuracy, errorbars = get_noise_accuracy(np.arange(0,100,5), np.arange(-10,10,0.01), i)
```

File: stratified vs. random cross-val.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+def get_noise_accuracy(noise_iterable,x_values,cut_off):
+    """Returns the noise iterable, the list of accuracy scores, and the list of standard deviations for error bars."""
+    bias = np.ones(len(x_values))
+    X = np.vstack([x_values,bias]) # Add Intercept
+    B =  [1., 1.] # Sigmoid params for X   
+    
+    mean_acc_scores = []
+    standard_deviations = []
+    noise_list = noise_iterable # Different noise values to be used in gaussian distribution sampling.
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]: I think this line is probably unnecessary - you should be able to just use `noise_list` directly as an argument to the function instead of `noise_iterable`.

File: stratified vs. random cross-val.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+
+# Creating a list that will be used for the different cutoffs in labeling.
+cut_off_list = [0.01,0.25,0.5,0.75,0.99]
+
+
+# Looping through cutoffs and graphing each model at each labeling breakpoint
+for i in cut_off_list:
+    noise,accuracy,errorbars = get_noise_accuracy(np.arange(0,100,5), np.arange(-10,10,0.01), i)
+    plt.errorbar(noise, accuracy, yerr = errorbars,label = str(i))
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]:
```suggestion
    plt.errorbar(noise, accuracy, yerr = errorbars, label = str(i))
```

File: stratified vs. random cross-val.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    plt.title('Noise vs. Accuracy')
+    plt.xlabel('Noise')
+    plt.ylabel('Accuracy')
+    plt.legend(title = 'Label Cut-off')
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]: I think you should be able to put these lines outside of the loop, since you only need to add the title and axis labels to the plot once:
```suggestion

plt.title('Noise vs. Accuracy')
plt.xlabel('Noise')
plt.ylabel('Accuracy')
plt.legend(title = 'Label Cut-off')
```

File: stratified vs. random cross-val.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+@author: matthewayala
+"""
+
+import numpy as np
+from sklearn.linear_model import LogisticRegression
+import matplotlib.pyplot as plt
+from sklearn.model_selection import StratifiedKFold
+from sklearn.model_selection import KFold
+
+# Simulate logistic data with/without noise
```
Comment (jjc2718) [inline stratified vs. random cross-val.py:15]: When I take code snippets from Stack Overflow or similar sites, I like to add a link to the original source as a comment. This is useful when someone else (or yourself, a few months later) looks at the code and wants to know what's going on.
```suggestion
# Simulate logistic data with/without noise
# https://stats.stackexchange.com/a/313962
```

File: stratified vs. random cross-val.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        transformed_p = [] # Empty list to hold new transformed data (1's and 0's)
+        for a in pnoisy_data:    
+            if a >= cut_off: # Turning values > 0.5 in 1's and < 0.5 into 0's
+                a = 1
+            else:
+                a = 0
+            transformed_p.append(a) # Adding transformed values to the empty list.
+        
+        # Turning labels into an array for stratifiedkfold
+        transformed_p = np.array(transformed_p)
```
Comment (jjc2718) [inline stratified vs. random cross-val.py:53]: You could do this in one line if you want to:
```suggestion
        transformed_p = (pnoisy_data >= cutoff).astype(int)
```
I'm not saying you have to make this change, though - if your version of the code is clearer to you, feel free to stick with it. In general, I think it's more important to be clear and obvious about what you're doing than to write clever one-liners, unless the one-liner is substantially faster than the longer code (which it probably isn't in this case, although it may be slightly faster for large arrays).

File: stratified vs. random cross-val.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+        # Creating Logistic regression instance
+        LogReg = LogisticRegression(max_iter=1000)
+        
+        # StratifiedKFold Splits
+        skf = StratifiedKFold(n_splits = 5)
+        # kf = KFold(n_splits = 5)
+        accuracy_scores = [] # Empty accuracy score list to be filled with each iteration of kfold
+        for train_index, test_index in skf.split(x_data,transformed_p):
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]: Instead of commenting out the call to `KFold` here, I'd suggest passing a bool argument to the function (call it `use_stratified_cv`), then choosing which split function to use here based on that argument:

```suggestion
        if use_stratified_cv:
            skf = StratifiedKFold(n_splits=5)
            split = skf.split(x_data, transformed_p)
        else:
            kf = KFold(n_splits=5)
            split = kf.split(x_data)
        accuracy_scores = [] # Empty accuracy score list to be filled with each iteration of kfold
        for train_index, test_index in split:
```

File: stratified vs. random cross-val.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+        
+        # Getting average of each k-fold section
+        mean_score = np.average(accuracy_scores)
+        mean_acc_scores.append(mean_score)
+        
+        # Getting standard deviation of each k-fold section
+        standard_deviation = np.std(accuracy_scores)
+        standard_deviations.append(standard_deviation)
+        
+    return noise_list,mean_acc_scores,standard_deviations
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]:
```suggestion
    return noise_list, mean_acc_scores, standard_deviations
```

File: stratified vs. random cross-val.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            if a >= cut_off: # Turning values > 0.5 in 1's and < 0.5 into 0's
+                a = 1
+            else:
+                a = 0
+            transformed_p.append(a) # Adding transformed values to the empty list.
+        
+        # Turning labels into an array for stratifiedkfold
+        transformed_p = np.array(transformed_p)
+        # Creating Logistic regression instance
+        LogReg = LogisticRegression(max_iter=1000)
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]: It might be cleaner to put parameters like `max_iter` and `n_splits` at the top of the file (or include them as arguments to the function), so that they're easier to find if you want to change them. So you'd have something like this at the top of the file:
```python
# model fitting parameters used in get_noise_accuracy
MAX_ITER = 1000
N_SPLITS = 5
```
then you'd pass them to the functions, e.g. 
```python
LogReg = LogisticRegression(max_iter=MAX_ITER)
```
This is especially important as your scripts get longer and your code gets split between multiple files - having all the parameters that affect the model in one place will make your life much easier.

File: stratified vs. random cross-val.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    mean_acc_scores = []
+    standard_deviations = []
+    noise_list = noise_iterable # Different noise values to be used in gaussian distribution sampling.
+    for i in noise_list:
+        # Creating simulated logistic data set with several noise values
+        pnoisy = logistic(X, B, noise=np.random.normal(loc=0., scale= i, size=len(x_values)))
+        
+        # Reshaping data for model    
+        x_data = x_values.reshape(len(x_values),1)
+        pnoisy_data = pnoisy.reshape(len(x_values),1)
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]:
```suggestion
        x_data = x_values.reshape(len(x_values), 1)
        pnoisy_data = pnoisy.reshape(len(x_values), 1)
```

File: stratified vs. random cross-val.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+
+# Simulate logistic data with/without noise
+
+def logistic(x, b, noise=None):
+    L = x.T.dot(b)
+    if noise is not None:
+        L = L+noise
+    return 1/(1+np.exp(-L))
+
+# Function that sumulates logistic data, trains and validates model on the data.
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]:
```suggestion
# Function that simulates logistic data, trains and validates model on the data.
```

File: stratified vs. random cross-val.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+                a = 0
+            transformed_p.append(a) # Adding transformed values to the empty list.
+        
+        # Turning labels into an array for stratifiedkfold
+        transformed_p = np.array(transformed_p)
+        # Creating Logistic regression instance
+        LogReg = LogisticRegression(max_iter=1000)
+        
+        # StratifiedKFold Splits
+        skf = StratifiedKFold(n_splits = 5)
```
Comment (jjc2718) [inline stratified vs. random cross-val.py]: You should specify a random seed here, so that the splits are reproducible:
```suggestion
        skf = StratifiedKFold(n_splits = 5, random_state=SEED)
```
Then you can add `SEED` as a parameter at the top of the file.

**Review by jjc2718 — COMMENTED** [review]
Nice work on this! This looks great for a first pass. I made some comments, feel free to start addressing them on your own if you want/have time, or we can discuss them tomorrow.

Some high-level comments:

* In general, it's best to try to create PRs that only change one distinct thing. In this case, that would probably mean one PR for the license file, and a separate PR for the Python script. For this first PR it's fine, but in the future as you get more used to Git and GitHub, we can work on limiting the scope of each of your PRs to a single distinct set of changes.

* We have [a code review checklist here](https://github.com/greenelab/onboarding/blob/master/extras/code_review_checklist.md), which might be a good place to start reading about our code style standards. Different organizations can have different standards (or sometimes no standard) for what they want code style to look like. I tried to give suggestions on some style things I noticed in your script here, but if you're up for it [installing a Python style linter](https://github.com/greenelab/onboarding/blob/master/extras/linter_install_tutorial.md) can be a good way to automate this process.

* For the filename of your script, I would usually try to avoid spaces - they can be messy for things like tab completion on the command line since they have to be escaped. I usually replace them with underscores, so for example a filename like `stratified_vs_random_cross_val.py` would be much better IMO.

File: stratified vs. random cross-val.py (hunk 5 above)
Comment (mayala1925) [reply stratified vs. random cross-val.py]: Awesome, I didn't even think that it would be adding those labels and and titles each run through the loop but that makes sense!

File: stratified vs. random cross-val.py (hunk 10 above)
Comment (mayala1925) [reply stratified vs. random cross-val.py]: That is a great tip.

File: stratified vs. random cross-val.py (hunk 13 above)
Comment (mayala1925) [reply stratified vs. random cross-val.py]: When I try and add this in I get this error - "Setting a random_state has no effect since shuffle is False. You should leave random_state to its default (None), or set shuffle=True."

I'm not quite sure where the shuffle parameter is? If it is a part of StratifiedKFold or not.

Comment (jjc2718) [reply stratified vs. random cross-val.py]: Ahh, I see. `shuffle` is a parameter to StratifiedKFold. I'd suggest setting `shuffle=True` and including the random seed.

That parameter tells StratifiedKFold to shuffle the data before splitting it into train and test sets. For your simulated data it doesn't matter that much because you're generating it in (essentially) a random order, but for real datasets the order might be meaningful: say, for example, you collected some cancer samples over time, and their order in the file is from earliest to most recent. You wouldn't want the model to just learn whether samples are old or new, since that pattern isn't likely to generalize well to other datasets. Randomizing (shuffling) the order would get rid of the old -> new pattern and allow your model to learn patterns in the data that are more likely to generalize.

So, in short - I'd generally recommend `shuffle=True`, both here and in the future.

Comment (mayala1925) [reply stratified vs. random cross-val.py]: Awesome, fixed it up!
